### PR TITLE
add command for uploading JavaScript file from TypeScript file

### PR DESCRIPTION
**Pull request by tineheller**

### CODE REVIEW

File: src/commands.ts [hunk 1]
```diff
@@ -13,6 +13,8 @@ import stripJsonComments = require('strip-json-comments');
 const urlExists = require('url-exists');
 // tslint:disable-next-line:no-var-requires
 const open = require('open');
+// tslint:disable-next-line:no-var-requires
+const tsc = require('typescript-compiler');
```
Comment (bkircher) [inline src/commands.ts]: This should be
```ts
import tsc = require('typescript-compiler');
```
With this, no warning is generated by tslint.

File: src/commands.ts [hunk 2]
```diff
@@ -408,32 +410,34 @@ export function viewDocumentation() {
...
+
+            _uploadScript(loginData, jsname).then((scriptname) => {
+                vscode.window.setStatusBarMessage('uploaded: ' + scriptname);
+                resolve();
+            }).catch((reason) => {
+                reject(reason);
+            });
+        }
+    });
+}
```
Comment (bkircher) [inline src/commands.ts]: You make the function `async` but you do not use `await` in the body but rather `Promise.then` and `Promise.catch`. This is a bit odd and looks inconsistent. Why not use `await`? Syntax is much easier to follow with `await`.

File: src/commands.ts [hunk 3]
```diff
@@ -408,32 +410,34 @@ export function viewDocumentation() {
...
-// }
+
+
+
+
+export async function uploadJSFromTS(loginData: nodeDoc.LoginData, textDocument: vscode.TextDocument): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
+
+        if (!textDocument || '.ts' !== path.extname(textDocument.fileName)) {
+            reject('No active ts script');
```
Comment (bkircher) [inline src/commands.ts]: Consider reject with an Error instance and immediately leave the function with an early return here. E.g.

```ts
return reject(new Error('No active ts script'));
```
See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/reject and https://stackoverflow.com/a/32536083/104274.

File: src/extension.ts [hunk 4]
```diff
@@ -191,6 +191,16 @@ export function activate(context: vscode.ExtensionContext): void {
         })
     );
 
+    // uploadJSFromTS
+    context.subscriptions.push(
+        vscode.commands.registerCommand('extension.vscode-janus-debug.uploadJSFromTS', (param) => {
+            if (vscode.window.activeTextEditor) {
+                const _param = vscode.window.activeTextEditor.document;
+                commands.uploadJSFromTS(loginData, _param);
+            }
```
Comment (bkircher) [inline src/extension.ts]: Having names beginning with an underline (e.g. `_param` or `_uploadScript`)  is not very idiomatic TS and is only used, if at all, for class properties AFAIK. You should use keywords/mechanisms like `public`, `private`, `export` to distinguish between API and non-API or to hide implementation details. What is the reason for those underlines?

File: src/commands.ts (hunk 1 above)
Comment (tineheller) [inline src/commands.ts]: Are you sure that line is correct? When I insert it, I get an error.

File: src/commands.ts (hunk 3 above)
Comment (tineheller) [inline src/commands.ts]: The reject is wrong, I have to show the error here.
On the subject of `new Error`. Yes, I will change all error `strings` to `Error` objects.

File: src/commands.ts (hunk 2 above)
Comment (tineheller) [inline src/commands.ts]: For now I've removed async. We should think of changing the extension, debugger, SDS module and everything to async/await. But please not now...

File: src/extension.ts (hunk 4 above)
Comment (tineheller) [inline src/extension.ts]: Yes I know I've removed the underlines, at least in this part of my source code.

**Comment by tineheller** [comment]
When you execute the command "Upload JS From TS", you won't see anything in the server console. You should get a JavaScript file in the path of your TypeScript file and you should see this JavaScript file in DOCUMENTS Manager under "scripts".

File: src/commands.ts (hunk 1 above)
Comment (bkircher) [inline src/commands.ts]: You're correct. I was wrong here

**Pull request merged**